Public/Src/Cache/DistributedCache.Host/Service/Internal/DistributedContentStoreFactory.cs

```diff
@@ -372,32 +369,51 @@ private List<string> GetAzureStorageSecretNames(StringBuilder errorBuilder)
             {
                 return null;
             }
+            var azureBlobStorageCredentialsKind = _distributedSettings.AzureBlobStorageUseSasTokens ? CredentialsKind.AzureBlobSASToken : CredentialsKind.AzureBlobPlainText;
+            retrieveSecretsRequests.AddRange(storageSecretNames.Select(secretName => new RetrieveSecretsRequest(secretName, azureBlobStorageCredentialsKind)));
 
-            if (_distributedSettings.EventHubSecretName != null &&
-                _distributedSettings.GlobalRedisSecretName != null)
+            if (string.IsNullOrEmpty(_distributedSettings.EventHubSecretName) ||
+                string.IsNullOrEmpty(_distributedSettings.GlobalRedisSecretName))
             {
-                var retryPolicy = CreateSecretsRetrievalRetryPolicy(_distributedSettings);
-                return await retryPolicy.ExecuteAsync(
-                    async () =>
-                    {
-                        return await _arguments.Host.RetrieveKeyVaultSecretsAsync(
-                            new List<string>(storageSecretNames)
-                            {
-                                _distributedSettings.EventHubSecretName,
-                                _distributedSettings.GlobalRedisSecretName,
-                                _distributedSettings.SecondaryGlobalRedisSecretName
-                            }.Where(s => s != null).ToList(),
-                            token);
-                    },
-                    token);
+                return null;
             }
 
-            return null;
+            retrieveSecretsRequests.Add(new RetrieveSecretsRequest(_distributedSettings.EventHubSecretName, CredentialsKind.EventHubPlainText));
+
+            retrieveSecretsRequests.Add(new RetrieveSecretsRequest(_distributedSettings.GlobalRedisSecretName, CredentialsKind.RedisPlainText));
+            if (!string.IsNullOrEmpty(_distributedSettings.SecondaryGlobalRedisSecretName))
+            {
+                retrieveSecretsRequests.Add(new RetrieveSecretsRequest(_distributedSettings.SecondaryGlobalRedisSecretName, CredentialsKind.RedisPlainText));
+            }
+
+            // Ask the host for credentials
+            var retryPolicy = CreateSecretsRetrievalRetryPolicy(_distributedSettings);
+            var secrets = await retryPolicy.ExecuteAsync(
+                async () => await _arguments.Host.RetrieveKeyVaultSecretsAsync(retrieveSecretsRequests, token),
+                token);
+            if (secrets == null)
+            {
+                return null;
+            }
+
+            // Validate requests match as expected
+            foreach (var request in retrieveSecretsRequests)
+            {
+                if (secrets.TryGetValue(request.Name, out var credential) && !credential.IsOfKind(request.Kind))
+                {
+                    throw new SecurityException($"The credentials produced by the host for secret named {request.Name} do not match the expected kind");

```

Comment:
I think this if block means that the secret was missing or the kind is different.

But the error message states only one case here.

I'm not sure how exactly your CB implementation would looks like, but DistributedService implementation can skip some secrets if they're unavailable.

Reply:
So maybe it make sense to separate two cases: the secret is missing, and the kind is different.

And I think that the first case is misconfiguration, but the second type of errors is actually a bug.

So we can use exceptions for the first case, but use `Contract.Assert` for the second one.

---
In reply to: [303561809](https://github.com/microsoft/BuildXL/pull/587#discussion_r303561809) [](ancestors = 303561809)

Reply:
The two cases are separated. You can only go inside the if block if the secret is present in the returned ones, and the kind does not match. For example, if you requested secret A of kind AzureBlobSAS, and it was found, but the actual kind was AzureBlobPAT, then this error would trigger.

The missing case is handled when the secret itself is actually requested, i.e. all calls to GetRequiredSecret

---
In reply to: [303562197](https://github.com/microsoft/BuildXL/pull/587#discussion_r303562197) [](ancestors = 303562197,303561809)